<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
      <link rel="shortcut icon" href="../img/favicon.ico" />
    <title>View SNPs in the GUI - Crossover Explorer</title>
    <link rel="stylesheet" href="../css/theme.css" />
    <link rel="stylesheet" href="../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
        <link href="../assets/_mkdocstrings.css" rel="stylesheet" />
        <link href="../css/xo.css" rel="stylesheet" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "View SNPs in the GUI";
        var mkdocs_page_input_path = "gui.md";
        var mkdocs_page_url = null;
      </script>
    
    <!--[if lt IE 9]>
      <script src="../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href=".." class="icon icon-home"> Crossover Explorer
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption"><span class="caption-text">Home</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="..">Getting Started</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Running the Applications</span></p>
              <ul class="current">
                  <li class="toctree-l1"><a class="reference internal" href="../peaks/">Peak Finder</a>
                  </li>
                  <li class="toctree-l1 current"><a class="reference internal current" href="#">View SNPs in the GUI</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#opening-the-gui">Opening the GUI</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#chromosome-display">Chromosome Display</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#filters">Filters</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#widgets">Widgets</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#block-size-widget">Block Size Widget</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#block-length-widget">Block Length Widget</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#minimum-coverage-widget">Minimum Coverage Widget</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#genome-match-widget">Genome Match Widget</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#how-filters-are-applied">How Filters are Applied</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#summary-plots">Summary Plots</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#chromosome-names">Chromosome Names</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#buttons">Buttons</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#exiting-the-gui">Exiting the GUI</a>
    </li>
    </ul>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../vis/">Generate Plots from the Command Line</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Developer Documentation</span></p>
              <ul>
                  <li class="toctree-l1"><a class="" href="../modules.md">None</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" >Widgets</a>
    <ul>
                <li class="toctree-l2"><a class="" href="../widgets.md">Widget Classes</a>
                </li>
                <li class="toctree-l2"><a class="" href="../filterbox.md">Filter Box</a>
                </li>
                <li class="toctree-l2"><a class="" href="../chromosomes.md">Chromosome Display</a>
                </li>
                <li class="toctree-l2"><a class="" href="../app.md">PeakViewerApp</a>
                </li>
    </ul>
                  </li>
                  <li class="toctree-l1"><a class="" href="../filters.md">Filters</a>
                  </li>
                  <li class="toctree-l1"><a class="" href="../main.md">None</a>
                  </li>
                  <li class="toctree-l1"><a class="" href="../tests.md">None</a>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="..">Crossover Explorer</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href=".." class="icon icon-home" aria-label="Docs"></a></li>
          <li class="breadcrumb-item">Running the Applications</li>
      <li class="breadcrumb-item active">View SNPs in the GUI</li>
    <li class="wy-breadcrumbs-aside">
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h1 id="view-snps-in-the-gui">View SNPs in the GUI</h1>
<p>The GUI allows users to explore the blocks generated by the <code>peaks</code> command.</p>
<p>The window has two parts, as shown by this screenshot:</p>
<p><img alt="screenshot" src="../images/gui_start.png" /></p>
<p>The left side of the screen displays a set of <strong>filters</strong> which allow you to set limits on sizes and other attributes of blocks.</p>
<p>The right side of the window is for graphical displays.  Note there are two tabs at the top of this part of the window.  The <strong>chromosome tab</strong>, shown above, displays one chromosome at a time.  Clicking the <strong>summary tab</strong> will change to show plots that summarize all the blocks in the data set.</p>
<h2 id="opening-the-gui">Opening the GUI</h2>
<p>From the command line run the <code>xo</code> script with the <code>gui</code> command:</p>
<pre><code class="language-bash">$ xo gui
</code></pre>
<p>The GUI is based on a Python library named Panel.  It's similar to Jupyter in that it creates a server on your local system, and you use the GUI by opening a web browser and connecting to the server.  The default port number is 5006, but you can specify a different port with the <code>--port</code> option if you need to (<em>e.g.</em> if you have a different web app already using that port).</p>
<p>The GUI needs two data files.  The first is the CSV file with block descriptions created by the <code>xo peaks</code> command.  The default file name is <code>peaks.csv</code>, but you can specify a different name using the <code>--peaks</code> option.</p>
<p>The other data file used by the GUI is a summary of locations of blocks of SNPs.  The default file name is <code>BSP_TIGER.intervals_dataframe.pickle.gzip</code>, but a different name can be specified using <code>--intervals</code>.</p>
<h2 id="chromosome-display">Chromosome Display</h2>
<p>At the top of the chromosome tab you will see a box containing a chromosome name with two buttons on either side.  Below that is a "ribbon" that classifies regions of the chromosome as having come from the CB4856 parent (in blue) or the N2 parent (in red).  The black dots represent the locations of blocks of SNPs identified by the peak finder.</p>
<p><img alt="chromosome" src="../images/chromosome.png" /></p>
<p>The controls allow you to switch to a different chromosome.  Click the triangle shaped buttons next to the chromosome name to move to the previous or next chromosome in the data set.  If you want to go directly do a different chromosome click on the chromosome name and edit the name there.  If the name is valid the display will be updated when you hit the enter key or click somewhere outside the text box.</p>
<blockquote>
<p>Note: If a chromosome has a lot of blocks there will be a delay before the screen is updated while the app makes all the graphics for the lower part of the display.</p>
</blockquote>
<p>Below the "ribbon" you will see the blocks identified by the peak finder for the current chromosome. This screenshot shows the first three blocks for chromosome BSP-OR-001-4:</p>
<p><img alt="dots" src="../images/blocks.png" /></p>
<p>The circles in each row correspond to the SNPs in that block.  The scale below the circles indicates the relative position of each SNP, and the text to the left has an overview of the block.  In this example, the first row of circles is for a block that starts 2Mbp into the chromosome.  The block has 19 SNPs, spread out over a region 421bp long (the distance between the first and last SNP in the block).</p>
<p>The colors of the circles are based on how the HMM in TIGER classified the SNPs.  </p>
<ul>
<li>Bright blue and red are where the HMM is confident a SNP is from CB4856 or N2, respectively.  </li>
<li>Paler blue and red are for SNPs with lower confidence.</li>
<li>Gray and yellow are for undetermined or ambiguous SNPs.</li>
</ul>
<p>To the left of each row of circles is a button with a "reveal" icon (looks like a greater-than symbol). If you click this button the GUI will show a table based on the data frame for the block, showing all the details of each SNP.  Here is the display for block #1 in the example:</p>
<p><img alt="table" src="../images/block_open.png" /></p>
<p>Note how there are 13 rows, one for each SNP, and that each row corresponds to a circle in the graphic.  Most of the circles are gray ("unknown") but one is bright red, having been identified as an N2 SNP.</p>
<p>Clicking the button again will hide the block.</p>
<h2 id="filters">Filters</h2>
<p>There are four ways to control which SNPs are displayed in the blocks:</p>
<ul>
<li>a <strong>block size filter</strong> limits the display to contain only blocks that have a number of SNPs between a lower and upper limit</li>
<li>a <strong>block length filter</strong> shows only blocks that have an overall length -- the distance, in base pairs, between the first and last SNP in the block -- within a specified lower and upper limit</li>
<li>the <strong>minimum coverage</strong> filter selects SNPs that are supported by reads (the values in the <code>ref_reads</code> and <code>var_reads</code> columns in the SNP table); a SNP is included in the display if the sum of the numbers in these two columns is greater then a specified limit</li>
<li>the <strong>genome match</strong> filter is either on or off; when it's on, a SNP is included only if the state inferred by the HMM (shown in the <code>hmm_state1</code> column in the table) matches the base in the chromosome (indicated by the <code>base_geno</code> column).</li>
</ul>
<h3 id="widgets">Widgets</h3>
<p>Each filter has its own "widget" on the left side of the GUI that is used to set values for that filter:</p>
<p><img alt="filters" src="../images/filters.png" /></p>
<h4 id="block-size-widget">Block Size Widget</h4>
<p>Click on the control on the left side of the slider and drag it to set the minimum block size, or click and drag the control on the right to set the maximum block size.</p>
<p>As you drag a control the display will update to show the new value.  The value won't take effect until you let go of the button.</p>
<h4 id="block-length-widget">Block Length Widget</h4>
<p>This widget is another double-sided slider, used the same way as the block size widget, but this one controls the block length.</p>
<h4 id="minimum-coverage-widget">Minimum Coverage Widget</h4>
<p>This slider has only one button, used to specify the coverage value.  You can either drag the button or click anywhere along the length of the slider.</p>
<h4 id="genome-match-widget">Genome Match Widget</h4>
<p>This widget is a simple check-box.  Click inside the box to turn the filter on or off.</p>
<h3 id="how-filters-are-applied">How Filters are Applied</h3>
<p>The chromosome display will be updated immediately, as soon you update a widget.</p>
<blockquote>
<p><strong>All four filters are applied each time the display is updated.</strong>
</p>
</blockquote>
<p>Each time the chromosome display is updated -- either by adjusting one of the widgets, or by selecting a new chromosome -- all the filters are applied and the resulting SNPs are used to generate the display.</p>
<ul>
<li>Only those blocks that are within the bounds of the block size and block length filters are shown.</li>
<li>Inside the blocks that pass those filters only SNPs that pass the coverage and match filters are shown.</li>
</ul>
<h2 id="summary-plots">Summary Plots</h2>
<p>The other tab in the graphics panel is named <strong>Summary</strong>.  Click on the tab name to display a set of controls for generating plots based on summarizing filtered blocks across all chromosomes:</p>
<p><img alt="summary" src="../images/summary.png" /></p>
<h3 id="chromosome-names">Chromosome Names</h3>
<p>The default setting will generate plots using all of the chromosomes.  However, if you want to make separate plots for different groups, you can enter a name pattern using regular expression syntax in the text box labeled "Chromosomes".</p>
<p>The initial value in this box is <code>BSP.*</code>, meaning "any chromosome with a name that starts with BSP" (in other words, all chromosomes).  Some other examples of name patterns:</p>
<table>
<thead>
<tr>
<th>pattern</th>
<th>chromosomes used</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>BSP-OR.*</code></td>
<td>all oocytes</td>
</tr>
<tr>
<td><code>BSP-SR.*</code></td>
<td>all spermatocytes</td>
</tr>
<tr>
<td><code>BSP-OR-10.*</code></td>
<td>the 10 worms with names BSP-OR-001, BSP-OR-002, ... BSP-OR-009.</td>
</tr>
<tr>
<td><code>BSP-SR-.*-1</code></td>
<td>chromosome 1 for all spermatocytes</td>
</tr>
</tbody>
</table>
<h3 id="buttons">Buttons</h3>
<p>The three summary plots are all histograms.  Click the button for the summary you want, based on block size, block length, or block location (relative location on the chromosome).</p>
<p>Note that the histograms will be based on filtered data.</p>
<ul>
<li>The application first selects the chromosomes that have names that match the pattern in the "Chromosomes" box.</li>
<li>It then applies all the filters, using the current settings in the filter widgets.</li>
<li>The result blocks are then used to generate the histograms.</li>
</ul>
<p>When the graph has been generated (it may take a few seconds) it will be displayed in the GUI:</p>
<p><img alt="histogram" src="../images/histogram.png" /></p>
<h2 id="exiting-the-gui">Exiting the GUI</h2>
<p>Close the web browser window.</p>
<p>If you started the GUI by typing <code>xo gui</code> on the command line you can stop it by clicking in the terminal window and typing <code>⌃C</code> (control-C).</p>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../peaks/" class="btn btn-neutral float-left" title="Peak Finder"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../vis/" class="btn btn-neutral float-right" title="Generate Plots from the Command Line">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
    
      <span><a href="../peaks/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../vis/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script src="../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "..";</script>
    <script src="../js/theme_extra.js"></script>
    <script src="../js/theme.js"></script>
      <script src="../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
