<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
      <link rel="shortcut icon" href="../img/favicon.ico" />
    <title>Crossover Explorer Documentation - Crossover Explorer</title>
    <link rel="stylesheet" href="../css/theme.css" />
    <link rel="stylesheet" href="../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
        <link href="../assets/_mkdocstrings.css" rel="stylesheet" />
        <link href="../css/xo.css" rel="stylesheet" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Crossover Explorer Documentation";
        var mkdocs_page_input_path = "vis.md";
        var mkdocs_page_url = null;
      </script>
    
    <!--[if lt IE 9]>
      <script src="../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href=".." class="icon icon-home"> Crossover Explorer
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption"><span class="caption-text">Home</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="..">Getting Started</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Running the Applications</span></p>
              <ul class="current">
                  <li class="toctree-l1"><a class="reference internal" href="../peaks/">Peak Finder</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../gui/">View SNPs in the GUI</a>
                  </li>
                  <li class="toctree-l1 current"><a class="reference internal current" href="#">Crossover Explorer Documentation</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#data-files">Data Files</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#use-peaks-to-find-interesting-blocks-of-snps">Use peaks to Find Interesting Blocks of SNPs</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#view-snps-in-the-gui">View SNPs in the GUI</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#example">Example</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#exiting-the-gui">Exiting the GUI</a>
    </li>
        </ul>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#visualizations">Visualizations</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#selecting-a-subset-of-the-data">Selecting a Subset of the Data</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#filter-options">Filter Options</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#example_1">Example</a>
    </li>
        </ul>
    </li>
    </ul>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Developer Documentation</span></p>
              <ul>
                  <li class="toctree-l1"><a class="" href="../modules.md">None</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" >Widgets</a>
    <ul>
                <li class="toctree-l2"><a class="" href="../widgets.md">Widget Classes</a>
                </li>
                <li class="toctree-l2"><a class="" href="../filterbox.md">Filter Box</a>
                </li>
                <li class="toctree-l2"><a class="" href="../chromosomes.md">Chromosome Display</a>
                </li>
                <li class="toctree-l2"><a class="" href="../app.md">PeakViewerApp</a>
                </li>
    </ul>
                  </li>
                  <li class="toctree-l1"><a class="" href="../filters.md">Filters</a>
                  </li>
                  <li class="toctree-l1"><a class="" href="../main.md">None</a>
                  </li>
                  <li class="toctree-l1"><a class="" href="../tests.md">None</a>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="..">Crossover Explorer</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href=".." class="icon icon-home" aria-label="Docs"></a></li>
          <li class="breadcrumb-item">Running the Applications</li>
      <li class="breadcrumb-item active">Crossover Explorer Documentation</li>
    <li class="wy-breadcrumbs-aside">
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h1 id="crossover-explorer-documentation">Crossover Explorer Documentation</h1>
<p>The <strong>crossover explorer</strong> is an application that filters and displays SNPs, with the goal of looking for evidence of crossover and non-crossover events in <em>C. elegans</em> genomes.</p>
<h2 id="data-files">Data Files</h2>
<p>The application uses two sets of data:</p>
<ul>
<li>
<p><code>BSP_TIGER.marker_dataframe.pickle.gzip</code> is a "pickled" and compressed Pandas data frame with the output from TIGER.  Each row in the frame describes a SNP, with columns for the
chromosome name, location, and the predicted parent genome (N2 or CB4856), and more.</p>
</li>
<li>
<p><code>BSP_TIGER.intervals_dataframe.pickle.gzip</code>, a summary of the SNP data, where each row defines a chromosome segment and its predicted parent.</p>
</li>
</ul>
<h2 id="use-peaks-to-find-interesting-blocks-of-snps">Use <code>peaks</code> to Find Interesting Blocks of SNPs</h2>
<p>To run the peak finder with the default options type</p>
<pre><code class="language-bash">$ xo peaks
</code></pre>
<p>That will read the SNP data from TIGER and write blocks to an output file in CSV format.</p>
<p>The options for this step allow you to choose different data file names for both the input and output file and to set an upper limit on the block size:</p>
<pre><code class="language-bash">$ xo peaks --help
usage: xo peaks [-h] [--snps F] [--output F] [--max_snps N]

options:
  -h, --help    show this help message and exit
  --snps F      input (IGER marker) file
  --output F    output file
  --max_snps N  max number of SNPs in a block
</code></pre>
<p>If <code>--max_snps</code> is not specified the default of 1000 SNPs per block is used.</p>
<h2 id="view-snps-in-the-gui">View SNPs in the GUI</h2>
<p>The <code>gui</code> command has options for specifying the data files to use and a network port used by the GUI:</p>
<pre><code class="language-bash">$ xo gui --help
usage: xo gui [-h] [--intervals F] [--peaks F] [--port N]

options:
  -h, --help     show this help message and exit
  --intervals F  SNP summaries
  --peaks F      blocks saved by peaks.py
  --port N       local port for the Panel server
</code></pre>
<p>Specify the path to the peak data with <code>--peaks</code> option; the default is <code>peaks.csv</code> (the default output name from the <code>xo peaks</code> command)</p>
<p>The other data file used by the GUI is a summary of locations of blocks of SNPs.  Specify the path to this file with <code>--intervals</code>; the default is <code>BSP_TIGER.intervals_dataframe.pickle.gzip</code></p>
<p>Panel, the Python library used to display the GUI, is similar to Jupyter.  It creates a server on your local system, and you use the GUI by opening a web browser and connecting to the server.  The default port number is 5006, but you can specify a different port number if you need to (<em>e.g.</em> if you have a different Panel app already using that port).</p>
<h3 id="example">Example</h3>
<p>The simplest workflow is to create a directory and copy (or link) the two pickled dataframes, using their default file names.</p>
<p>Run the <code>peaks</code> command with default options:</p>
<pre><code class="language-bash">$ xo peaks
</code></pre>
<p>It will take a few minutes, but the script will show a status line to indicate it's still working.  When it's done there will be a file named <code>peaks.csv</code> in the current directory.</p>
<p>Run the <code>gui</code> command:</p>
<pre><code class="language-bash">$ xo gui
loading interval data
loading peak data
Launching server at http://localhost:5006
</code></pre>
<p>That will read the two data files and start the GUI, which uses a web server connected to port 5606.  The server should start your default web browser and show a view of the first chromosome in the data set.</p>
<p>If you browser doesn't start automatically, just start the browser, open a new window, and enter the URL printed in the terminal window (<code>http://localhost:5006</code> in the example above).</p>
<h4 id="exiting-the-gui">Exiting the GUI</h4>
<p>Close the web browser window, and type <code>^C</code> in the terminal window where you typed the <code>xo view</code> command.</p>
<h2 id="visualizations">Visualizations</h2>
<p>The visualization command will read the data, apply the same filters available in the GUI, and display the resulting graphic in a window.</p>
<p>Currently there are three different visualizations:</p>
<ul>
<li>a histogram of the number of SNPs in each block</li>
<li>a histogram of the length (in base pairs) of each block</li>
<li>a histogram of the relative location on the chromosome of each block</li>
</ul>
<p>When running the command type <code>xo vis P</code> where <code>P</code> is the type of plot to make, either <code>count</code>, <code>length</code>, or <code>location</code>.</p>
<p>The remaining options are:</p>
<pre><code class="language-bash">  --peaks F        blocks saved by peaks.py
  --chromosomes P  names of chromosomes to use
  --size N N       block size range (#SNPs)
  --length N N     block length range (bp)
  --coverage N     minimum coverage
  --match          require genome match
</code></pre>
<p>The <code>--peaks</code> option specifies the name of the peak data file (same as in the <code>gui</code> command).</p>
<h3 id="selecting-a-subset-of-the-data">Selecting a Subset of the Data</h3>
<p>The <code>--chromosome</code> option lets you select a subset of the data to visualize by specifying a regular expression to match the chromosome name.  The default is <code>BSP.*</code>, meaning "any chromosome with a name that starts with BSP" (in other words, all chromosomes).  Some other examples of name patterns:</p>
<table>
<thead>
<tr>
<th>pattern</th>
<th>chromosomes used</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>BSP-OR.*</code></td>
<td>all oocytes</td>
</tr>
<tr>
<td><code>BSP-SR.*</code></td>
<td>all spermatocytes</td>
</tr>
<tr>
<td><code>BSP-OR-10.*</code></td>
<td>the chromosomes for the 10 worms with names BSP-OR-001, BSP-OR-002, ... BSP-OR-009.</td>
</tr>
<tr>
<td><code>BSP-SR-.*-1</code></td>
<td>chromosome 1 for all spermatocytes</td>
</tr>
</tbody>
</table>
<h3 id="filter-options">Filter Options</h3>
<p>The remaining options allow you to specify the filters to use.  There is one option for each of the filters shown in the GUI, so the plots that are made by this command are accurate summaries based on settings you make in the GUI.</p>
<p>Two of the options (<code>--size</code> and <code>--length</code>) correspond to the interval selectors and require you to specify two values, the minimum and maximum slider locations.  If you don't specify these options the defaults displayed in the GUI are used: block size from 0 to 100 SNPs, block length from 0 to 10,000 bp.</p>
<p>The <code>--coverage</code> option is a single integer based on the location of that slider in the GUI.  The default is 0.</p>
<p>Use <code>--match</code> option if you clicked the checkbox labeled Genome Match.</p>
<h3 id="example_1">Example</h3>
<p>This command generates a histogram of block sizes (number of SNPs) using chromosome 1 from all worms, restricted to SNPs that have a genome match and using only blocks between 1 and 1000 bp long:</p>
<pre><code class="language-bash">$ xo vis count --chr 'BSP-.*-1' --match --length 0 1000
</code></pre>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../gui/" class="btn btn-neutral float-left" title="View SNPs in the GUI"><span class="icon icon-circle-arrow-left"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
    
      <span><a href="../gui/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
  </span>
</div>
    <script src="../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "..";</script>
    <script src="../js/theme_extra.js"></script>
    <script src="../js/theme.js"></script>
      <script src="../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
